openpgp/ocb/ocb.go

```diff
@@ -0,0 +1,269 @@
+
+import (
+	"math/bits"
+	"bytes"
+	"crypto/cipher"
+	"golang.org/x/crypto/openpgp/bytesutils"
+)
+
+type ocb struct {
+	block     cipher.Block
+	tagSize   int
+	nonceSize int
+	mask      mask
+}
+
+type mask struct {
+	// L_*, L_$, (L_i)_{i ∈ N}
+	L_ast []byte
+	L_dol []byte
+	L     [][]byte
+}
+
+const (
+	defaultTagSize     = 16
+	defaultNonceSize   = 15
+)
+
+// NewOCB returns an OCB instance with the given block cipher and default
+// tag and nonce sizes.
+func NewOCB(block cipher.Block) ocb {
+
+	return ocb{
+		block:     block,
+		tagSize:   defaultTagSize,
+		nonceSize: defaultNonceSize,
+		mask:      initializeMaskTable(block),
+	}
+}
+
+// NewOCBWithNonceAndTagSize returns an OCB instance with the given block
+// cipher, nonce length, and tag length. Panics on zero nonceSize and
+// exceedingly long tag size.
+//
+// It is recommended to use at least 12 bytes as tag length.
+func NewOCBWithNonceAndTagSize(block cipher.Block, nonceSize, tagSize int) ocb {
+	if nonceSize == 0 {
+		panic(`Error: Can't initialize OCB instance with nonceSize = 0`)
+	}
+	if nonceSize >= block.BlockSize() {
+		panic(`Error: Custom nonce length exceeds blocksize - 1`)
+	}
+	if tagSize > block.BlockSize() {
+		panic(`Error: Custom tag length exceeds blocksize`)
+	}
+	return ocb{
+		block:     block,
+		tagSize:   tagSize,
+		nonceSize: nonceSize,
+		mask:      initializeMaskTable(block),
+	}
+}
+
+func (o *ocb) NonceSize() int {
+	return o.nonceSize
+}
+
+func (o *ocb) TagSize() int {
+	return o.tagSize
+}
+
+// Encrypt returns a byte array containing the concatenation of the
+// ciphertext and the validation tag.
+func (o *ocb) Encrypt(plaintext, nonce, adata []byte) []byte {
+	return o.crypt("enc", nonce, adata, plaintext)
+}
+
+// Decrypt returns a byte array containing the plaintext. If the input tag
+// (contained at the end of the ciphertext) is invalid, it panics.
+func (o *ocb) Decrypt(ciphertext, nonce, adata []byte) []byte {
+	sep := len(ciphertext)-o.tagSize
+	ciphertextData := ciphertext[:sep]
+	tag := ciphertext[sep:]
+	plaintextAndTag := o.crypt("dec", nonce, adata, ciphertextData)
+	if bytes.Equal(tag, plaintextAndTag[sep:]) {
+		return plaintextAndTag[:sep]
+	}
+	panic("Authentication tag mismatch")
+}
+
+// On instruction == "enc" (resp. "dec"), crypt is the encrypt (resp. decrypt)
+// function. It returns the resulting plain/ciphertext with the tag appended.
+func (o *ocb) crypt(instruction string, nonce, adata, text []byte) []byte {
+	//
+	// Consider X as a sequence of 128-bit blocks
+	//
+	// Note: For encryption X is the plaintext, for decryption X is the
+	// ciphertext without the tag.
+	X := make([]byte, len(text))
+	copy(X, text)
+	blockSize := o.block.BlockSize()
+
+
+	//
+	// Nonce-dependent and per-encryption variables
+	//
+	// From RFC 7253:
+	// Nonce = num2str(TAGLEN mod 128, 7) || zeros(120 - bitlen(N)) || 1 || N
+	paddedNonce := append(make([]byte, blockSize-1-len(nonce)), 1)
+	paddedNonce = append(paddedNonce, nonce...)
```

Comment:
I'd love to, but `append` only takes two arguments...

Reply:
I'm pretty sure you can append multiple elements, since the second argument is variadic (indicated by the `...` in the definition of the function). In fact, what you're technically doing here is also passing multiple arguments, it's just that the arguments come from a list (indicated by the `...` in the function call). You can combine the two, passing some arguments directly and the remainder from a list.

Reply:
I understand the reasoning, but it still doesn't seem to be the correct syntax. For instance, on this following example, it says `too many arguments to append`:
 
```
package main

func main() {
	a := []int{1,2,3}
	b := []int{4,5,6}
	c := []int{7,8,9}
	d := append(a, b, c...)
}
```

Reply:
Interesting, you're right. Apologies ^.^ Go is weird.

Reply:
Yes, it's odd that append is not variadic, yet `fmt.Println` is...